src/vs/workbench/browser/parts/editor/editorGroupView.ts

```diff
@@ -1602,7 +1602,7 @@ export class EditorGroupView extends Themable implements IEditorGroupView {
 		let activeReplacement: EditorReplacement | undefined;
 		const inactiveReplacements: EditorReplacement[] = [];
 		editors.forEach(({ editor, replacement, options }) => {
-			if (editor.isDirty() && !editor.isSaving()) {
+			if (!editor.isUntitled() && editor.isDirty() && !editor.isSaving()) {
```

Comment:
This looks scary to me, why? Keep in mind we might be calling `replaceEditors` from this flow:
* new untitled
* make dirty
* file > save
* it should replace the untitled with the saved editor

Reply:
Maybe I'm confused about the concern here. The flow you describe does work.

Reply:
Maybe can you enlighten me why this change is needed here?

Reply:
This change is needed as currently if you open a custom editor / notebook using `Reopen With...` then the untitled tab doesn't get replaced because it is a dirty untitled file. This API allows you to reopen untitled files as custom editors / notebooks as the data is transferred to the custom editor / notebook therefore it is safe to dispose of the the untitled editor.

Reply:
I think the `replaceEditors` method should take care of dirty editors by asking the user to save or revert and then this check can go away entirely. Will push this today.

Reply:
This is pushed via https://github.com/microsoft/vscode/commit/f294b6f81090546583ba995b947d66ed468f8d17

Reply:
We don't want it to do that for untitled files as then it breaks the workflow of untitled files in custom editors. Like there is no way for me to have an untitled file do some work and then want to view it in a different editor as the replacing will force me to save which means my file is no longer untitled.

Reply:
I think in that case `replaceEditors` is not the right method to use for this flow which seems to want to do this:
* persist the contents of a text editor in memory
* close an editor even if dirty (we don't have code for this for obvious reasons because it is essentially data loss)
* at that location open a new custom editor with the data

Maybe the editor group needs to be aware of such an operation, or it is implemented outside, but then we possibly end up with something like we have now (`editorOpenWith.ts`), which I am not a big fan of...

Reply:
It is a unique case in where a dirty file being closed isn't data loss as this PR persists that state into the custom editor. What about adding a flag to replaceEditors letting it force a replace?

Reply:
Yeah having a very explicit flag for this in `replaceEditors` sounds more reasonable to me so that this is not the default behaviour but needs handling from whoever calls this method.

Reply:
Btw this would not be specific to untitled editors, but also files right? And even custom editors?

Reply:
No this would be specific to untitled editors that are transitioning to custom editors / notebooks. For files we want you to save as the changes can often be quite large so we don't want to push that over to the extension host and we would prefer to give the custom editor the URI to do with as they please. The issue with untitled files is you can't really do anything with that URI as it doesn't resolve to an actual filesystem and therefore we give you the bytes.

Reply:
What would happen when I reopen a dirty file (not untitled) with a custom editor then? You just save?

Reply:
Correct it should ask you to save as it's already a file on the file system so it's not a huge deal to ask you to save or discard your changes so that the custom editor can use the fs APIs. The reason here we don't want you to save is the whole point of having an untitled file is to have this idea of a "temp" file.